modules/iap-tunneling/main.tf

```diff
@@ -0,0 +1,45 @@
+/**
+ * Copyright 2018 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+resource "google_compute_firewall" "allow_from_iap_to_instances" {
+  project = var.host_project != "" ? var.host_project : var.project
+  name    = var.fw_name_allow_ssh_from_iap
+  network = var.network
+
+  allow {
+    protocol = "tcp"
+    ports    = ["22"]
+  }
+
+  # https://cloud.google.com/iap/docs/using-tcp-forwarding#before_you_begin
+  # This is the netblock needed to forward to the instances
+  source_ranges = ["35.235.240.0/20"]
+
+  # Use service account instead of network tags to identify the instance to be more restrict and explict.
+  # https://cloud.google.com/vpc/docs/firewalls#service-accounts-vs-tags
+  target_service_accounts = [var.service_account]
```

Comment:
I didn't include that because using tags would open up access to other VMs sharing the same tags, which might not a good idea. And people with wider range of permissions can modify tags. (https://cloud.google.com/vpc/docs/firewalls#service-accounts-vs-tags)

Since the iam binding part applies to only one VM, ideally, I want to be able to just specify the VM instance in the firewall rule as well. But I can't. So using the service account associated with that VM can serve the similar purpose.

We could add support for tags if we want to enable IAP for multiple VMs at once in this module, but that could get a little bit messy in the iam binding part.

Reply:
I completely agree that tags are not as ideal as service accounts, but there are uses for them and many folks use them, so for something that may not be critical for security, such as a bastion, I think we should add them in, if this happens in a v2, that's fine but we should track it in an issue after this is merged.

Also I didn't notice upon first review that the target is a single instance. We should definintely allow passing in a list of instances, otherwise users will have to duplicate this module for N instances, which will not work since this module also creates firewall rules that have to have unique names. This needs to be a v1 thing. I'll add an inline comment there so it doesn't get lost.

Reply:
Added support network tags as well as making instances a list.